### PR TITLE
Invoice Auto-Deletion and Allow Clients to Delete Unpaid Old Invoices

### PR DESCRIPTION
**Pull request by sowmiyasubramaniyan**
Run: php artisan migrate
Implemented changes as discussed in #2589 
Auto-deletion run the below command:
php artisan invoices:delete

### CODE REVIEW

File: app/Http/Controllers/Common/CronController.php
Comment (sandesh556) [inline app/Http/Controllers/Common/CronController.php:451]: Don't nest your loop inside the codition, if(false) just return and then if it  is true it will execute your code

File: app/Http/Controllers/Common/CronController.php
Comment (sandesh556) [inline app/Http/Controllers/Common/CronController.php:493]: DB::transaction(function () use ($invoice) {
    $invoice->invoiceItem()->delete();
    
    if ($invoice->is_renewed != 0 && $invoice->orderRelation()->exists()) {
        $invoice->orderRelation()->delete();
    }
    
    $invoice->delete();
});

File: app/Http/Controllers/Common/CronController.php
Comment (sandesh556) [inline app/Http/Controllers/Common/CronController.php:471]: Could also do this $oldInvoices = Invoice::where('status', 'pending')
                      ->where('date', '<', $date)
                      ->with(['invoiceItem', 'orderRelation'])
                      ->get();

File: app/Http/Controllers/Front/CartController.php
Comment (sandesh556) [inline app/Http/Controllers/Front/CartController.php:181]: Put this in one private method

File: app/Http/Controllers/Front/CartController.php
Comment (sandesh556) [inline app/Http/Controllers/Front/CartController.php:188]: Put this in another private method

File: app/Http/Controllers/Front/ClientController.php
Comment (sandesh556) [inline app/Http/Controllers/Front/ClientController.php:902]: public function invoiceDelete($id)
{
    $invoice = Invoice::find($id);

    if (! $invoice) {
        return response()->json(['error' => 'Invoice not found'], 404);
    }

    if ($this->canDeleteInvoice($invoice)) {
        $this->deleteInvoice($invoice);

        return response()->json(['message' => 'Invoice deleted successfully']);
    }

    return response()->json(['error' => 'Cannot delete invoice.'], 400);
}

private function canDeleteInvoice($invoice)
{
    return (
        $invoice->is_renewed == 0 &&
        ! $invoice->orderRelation()->exists() &&
        $invoice->invoiceItem()->exists()
    ) || (
        $invoice->is_renewed != 0 &&
        $invoice->orderRelation()->exists() &&
        $invoice->invoiceItem()->exists()
    );
}

private function deleteInvoice($invoice)
{
    $invoice->invoiceItem()->delete();

    if ($invoice->is_renewed != 0 && $invoice->orderRelation()->exists()) {
        $invoice->orderRelation()->delete();
    }

    $invoice->delete();
    \Session::forget('invoice');
}
 

This can be split like this

File: app/Http/Controllers/Order/RenewController.php
Comment (sandesh556) [inline app/Http/Controllers/Order/RenewController.php:308]: $subscription = Subscription::find($id);

$latestInvoiceItem = InvoiceItem::where('plan_id', $request->plan)
    ->whereHas('invoice', function ($query) {
        $query->where('is_renewed', 1)->where('status', 'pending');
    })
    ->latest('created_at')
    ->first();

if ($latestInvoiceItem) {
    return redirect()->route('my-invoice', ['id' => $latestInvoiceItem->invoice_id])
        ->with('warning', 'You have an existing unpaid invoice for this product. Please proceed with the payment or delete the invoice and try again');
}

Move this also to a different private method

**Review by sandesh556 — CHANGES_REQUESTED** [review]

**Review by RafficMohammed — DISMISSED** [review]

File: app/Http/Controllers/Front/CartController.php
Comment (sandesh556) [inline app/Http/Controllers/Front/CartController.php:197]: Too much code just do this 

 $userId = auth()->id();
    $itemName = $item->name;
    $itemQuantity = $item->quantity;

    return Invoice::where('user_id', $userId)
        ->where('is_renewed', 0)
        ->where('status', 'pending')
        ->whereHas('invoiceItem', function ($query) use ($itemName, $itemQuantity) {
            $query->where('product_name', $itemName)->where('quantity', $itemQuantity);
        })
        ->firstOr(function () use ($item) {
            Cart::clear($item->id);
        });

File: app/Http/Controllers/Common/CronController.php
Comment (sandesh556) [inline app/Http/Controllers/Common/CronController.php:470]: Lets discuss this method

File: app/Http/Controllers/Common/CronController.php
Comment (sandesh556) [inline app/Http/Controllers/Common/CronController.php:481]: Lets discuss this method

File: app/Http/Controllers/Common/CronController.php
Comment (sandesh556) [inline app/Http/Controllers/Common/CronController.php:494]: Lets discuss all of this

**Review by sandesh556 — CHANGES_REQUESTED** [review]

**Review by sandesh556 — APPROVED** [review]

**Review by SarfrazS — CHANGES_REQUESTED** [review]
- [ ] Issue:- When we select stripe and razorpay and click on proceed getting below error.
image

**Comment by SarfrazS** [comment]
The developer told us once the client theme pr merge in the development branch after that you can test this PR so I'm putting it on hold now.

**Comment by ganthimathin** [comment]
@sowmiyasubramaniyan: Please resolve the conflicts.

**Review by ganthimathin — CHANGES_REQUESTED** [review]
Verified the functionality in local instance with existing DB. Found below issues:

1. The warning message 'You have an unpaid invoice for this product. Please proceed with the payment or delete the invoice and try again' is showing for freelancer product also.
2. When trying to change the cron settings as year for ex. and click update button exception throws. After that unable to pay for the product.

3.For freelance product, if trying to click pay from invoice page- alignment is totally changed.

4.When trying to Renew the order - 'Cannot view invoice. Invalid modification of data.' message is showing. Is this correct?

renew error message
while trying to save the cron settings exception throws
payment page for freelancer format is changed
for freelancer product also this error message is showing

**Comment by sowmiyasubramaniyan** [comment]
I've been unable to replicate this issue and have also discussed it with @ganthimathi. @SarfrazS, could you please cross-check these issues?

**Pull request merged**
